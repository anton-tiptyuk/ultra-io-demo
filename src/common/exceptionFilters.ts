import {
  ArgumentsHost, Catch, ExceptionFilter,
  HttpException,
} from '@nestjs/common';

import { QueryFailedError } from 'typeorm';
import { EntityNotFoundError } from 'typeorm/error/EntityNotFoundError';

const ERROR_CODE_ENTITY_NOT_FOUND = 'ERROR_CODE_ENTITY_NOT_FOUND';
const ERROR_CODE_FOREIGN_KEY = 'ERROR_CODE_FOREIGN_KEY';
const ERROR_CODE_DUPLICATE_UNIQUE = 'ERROR_CODE_DUPLICATE_UNIQUE';

const QUERY_FAILED_ERROR_CODE_F_KEY = '23503';
const QUERY_FAILED_ERROR_CODE_DUPLICATE_UNIQUE = '23505';

interface IResponse {
  message: string;
  details?: string;
  error: string;
  timestamp: string;
  path: string;
}

const referencedMessage = (entity: string, parent: string): string =>
  parent
    ?
    `Can't delete the "${entity}" that is used in the "${parent}"`
    :
    `Cannot set reference in ${entity} - foreign constraint violated`;

class BaseExceptionFilter implements ExceptionFilter {
  getResponse(defaultResponse: IResponse, exception: Error): object {
    return defaultResponse;
  }

  getCustomStatus(exception: Error) {
    return 500;
  }

  catch(exception: Error, host: ArgumentsHost) {
    const ctx = host.switchToHttp();
    const res = ctx.getResponse();
    const req = ctx.getRequest();

    const { message, name } = exception;

    const defaultResponse: IResponse = {
      message,
      error: name,
      timestamp: new Date().toISOString(),
      path: req.url,
    };

    console.log('BaseExceptionFilter: exception caught');
    console.log(defaultResponse);
    exception.stack && console.log(exception.stack);

    const response = this.getResponse(
      defaultResponse,
      exception
    );

    res
      .status(this.getCustomStatus(exception))
      .json(response);
  }
}

@Catch(HttpException)
class HttpExceptionFilter extends BaseExceptionFilter {
  getCustomStatus(exception: Error) {
    return (<HttpException>exception).getStatus();
  }

  getResponse(defaultResponse: IResponse, exception: Error) {
    const { timestamp, path } = defaultResponse;
    return Object.assign(
      (<HttpException>exception).getResponse(),
      { timestamp, path }
    );
  }
}

@Catch(EntityNotFoundError)
class EntityNotFoundFilter extends BaseExceptionFilter {

  getCustomStatus(exception: Error) {
    return 404;
  }

  getResponse(defaultResponse: IResponse, exception: Error) {
    return Object.assign(
      defaultResponse,
      { errorCode: ERROR_CODE_ENTITY_NOT_FOUND },
    );
  }
}

@Catch(QueryFailedError)
class QueryFailedErrorFilter extends BaseExceptionFilter {

  getCustomStatus(exception: Error) {
    return 400;
  }

  getResponse(defaultResponse: IResponse, exception: Error) {
    const code = exception['code'];
    const result: any = {};

    switch (code) {
      case QUERY_FAILED_ERROR_CODE_F_KEY:
        const [entity, , parent] = exception.message.match(/"(\w*)"/g);
        result.message = referencedMessage(entity, parent);
        result.errorCode = ERROR_CODE_FOREIGN_KEY;
        break;
      case QUERY_FAILED_ERROR_CODE_DUPLICATE_UNIQUE:
        result.errorCode = ERROR_CODE_DUPLICATE_UNIQUE;
        break;
    }

    return Object.assign(defaultResponse, result);
  }
}

@Catch()
class InternalServerErrorFilter extends BaseExceptionFilter { }

export const getFilters = () =>
  [
    new InternalServerErrorFilter(),
    new HttpExceptionFilter(),
    new EntityNotFoundFilter(),
    new QueryFailedErrorFilter(),
  ];
