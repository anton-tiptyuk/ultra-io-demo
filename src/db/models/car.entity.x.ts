import {
  Entity,
  PrimaryGeneratedColumn,
  Column,
  CreateDateColumn,
  UpdateDateColumn,
  ManyToOne,
  OneToMany,
  Index,
  BeforeUpdate,
} from 'typeorm';

// import { Exclude } from 'class-transformer';
// import { BoxToUser } from './boxToUser.entity';

@Entity()
// @Index('uq_box_title_lower', { synchronize: false })
export class Car {
  @PrimaryGeneratedColumn() id: number;

  // @Column({ length: 100 })
  // title: string;

  // @Column({ length: 100, unique: true })
  // barcode: string;

  // @Column({ length: 5 })
  // currency: string;

  // @ManyToOne((type) => PayDestination, { onDelete: 'CASCADE', nullable: true })
  // payDestination: PayDestination;

  // @Column()
  // allowCash: boolean;

  // @Column()
  // allowCashless: boolean;

  // @Exclude()
  // @CreateDateColumn()
  // createdAt: Date;

  // @Exclude()
  // @UpdateDateColumn()
  // updatedAt: Date;

  // @Exclude()
  // @OneToMany((type) => BoxToUser, boxToUser => boxToUser.box)
  // boxToUsers: BoxToUser[];

  // private checkPaymentStuff() {
  //   if (this.payDestination && !this.allowCashless)
  //     throw new BadRequestException('You should enable cashless payments to set payment destination');
  //   if (this.allowCashless && !this.payDestination)
  //     throw new BadRequestException('You should set payment destination to enable cashless payments');
  //   if (!(this.allowCash || this.allowCashless))
  //     throw new BadRequestException('The box should have at least one payment method enabled');
  // }

  // @BeforeUpdate()
  // beforeUpdate() {
  //   this.checkPaymentStuff();
  // }

}
