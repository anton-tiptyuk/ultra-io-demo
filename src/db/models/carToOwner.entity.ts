import {
  Entity,
  PrimaryGeneratedColumn,
  ManyToOne,
  Column,
} from 'typeorm';

import { Car } from './car.entity';
import { Owner } from './owner.entity';

@Entity()
export class CarToOwner {
  @PrimaryGeneratedColumn()
  id: number;

  @ManyToOne(() => Car, car => car.carToOwners, { onDelete: 'CASCADE' })
  car: Car;

  @ManyToOne(() => Owner, owner => owner.carToOwners, { onDelete: 'CASCADE' })
  owner: Owner;

  @Column()
  purchaseDate: Date;
}
