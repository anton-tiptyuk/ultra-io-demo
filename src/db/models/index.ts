export * from './owner.entity';
export * from './manufacturer.entity';
export * from './car.entity';