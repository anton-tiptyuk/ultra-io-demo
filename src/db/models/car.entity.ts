import {
  Entity,
  PrimaryColumn,
  Column,
  ManyToOne,
  OneToMany,
} from 'typeorm';

import { decimalTransformer } from '../../common/util';

import { Manufacturer } from './manufacturer.entity';
import { CarToOwner } from './carToOwner.entity';

@Entity()
// @Index('uq_box_title_lower', { synchronize: false })
export class Car {
  @PrimaryColumn({ length: 40 })
  id: string;

  @ManyToOne(() => Manufacturer, { onDelete: 'CASCADE', nullable: false })
  manufacturer: Manufacturer;

  @Column('decimal', { precision: 10, scale: 2, transformer: decimalTransformer })
  price: number;

  @Column()
  firstRegistrationDate: Date;

  @OneToMany(() => CarToOwner, carToOwners => carToOwners.car)
  carToOwners: CarToOwner[];
}
