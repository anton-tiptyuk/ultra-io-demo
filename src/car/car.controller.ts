import { Controller, Get, Param } from '@nestjs/common';
import { Crud, CrudController } from '@nestjsx/crud';

import { Car } from '../db/models';
import { CarService } from './car.service';

@Crud({
  model: { type: Car },
  query: {
    join: {
      manufacturer: { eager: true },
    },
  },
  params: {
    id: {
      // type: 'uuid',
      // though i seed UIDs for all entitites, the task says ID fields should be string
      type: 'string',
      primary: true,
      field: 'id',
    },
  },
  routes: { exclude: ['createManyBase', 'replaceOneBase'] },
})
@Controller('car')
export class CarController implements CrudController<Car> {
  constructor(public service: CarService) { }

  @Get(':id/manufacturer')
  getManufacturer(@Param('id') id: string) {
    return this.service.getManufacturer(id);
  }
}
