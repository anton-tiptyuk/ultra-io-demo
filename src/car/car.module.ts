import { Module } from '@nestjs/common';
import { TypeOrmModule } from '@nestjs/typeorm';

import { CarController } from './car.controller';
import { CarService } from './car.service';
import { Car, CarToOwner } from '../db/models';

@Module({
  imports: [TypeOrmModule.forFeature([Car, CarToOwner])],
  controllers: [CarController],
  providers: [CarService],
  exports: [CarService],
})
export class CarModule { }
