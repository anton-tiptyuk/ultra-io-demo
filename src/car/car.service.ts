import { Injectable } from '@nestjs/common';
import { Repository } from 'typeorm';
import { InjectRepository } from '@nestjs/typeorm';
import { TypeOrmCrudService } from '@nestjsx/crud-typeorm';

import { Car } from '../db/models';

@Injectable()
export class CarService extends TypeOrmCrudService<Car> {
  constructor(@InjectRepository(Car) repo: Repository<Car>) {
    super(repo);
  }

  decidePagination() {
    return false;
  }
}
