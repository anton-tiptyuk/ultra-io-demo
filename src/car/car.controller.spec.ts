import { Test } from '@nestjs/testing';
import { INestApplication } from '@nestjs/common';

import * as supertest from 'supertest';

import { Car } from '../db/models';
import { CarController } from './car.controller';
import { CarService } from './car.service';

const carServiceMock = {
  getManufacturer: () => undefined,
  createOne: () => undefined,
  getOne: () => undefined,
};

const spies = {
  carService: {
    getManufacturer: jest.spyOn(carServiceMock, 'getManufacturer'),
    createOne: jest.spyOn(carServiceMock, 'createOne'),
    getOne: jest.spyOn(carServiceMock, 'getOne'),
  },
};

let app: INestApplication;
const request = () => supertest(app.getHttpServer());

describe('Car Controller', () => {
  beforeAll(async () => {
    const module = await Test.createTestingModule({
      providers: [
        {
          provide: CarService,
          useValue: carServiceMock,
        }
      ],
      controllers: [CarController],
    }).compile();

    app = await module.createNestApplication().init();
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  it('should be defined', () => {
    expect(app.get(CarController)).toBeDefined();
  });

  it('should fetch manufacturer', async () => {
    const resultMock = { value: 'anything' };
    spies.carService.getManufacturer.mockImplementation(async () => resultMock);
    const { body } = await request().get('/car/random-id-value/manufacturer');
    expect(body).toEqual(resultMock);
    expect(spies.carService.getManufacturer).toHaveBeenCalledWith('random-id-value');
  });

  describe('checks for CRUD POST /car', () => {
    const createOneResult = { dummy: 'result' };

    const subject = (data: Object) => request().post('/car').send(data);

    beforeEach(() => {
      spies.carService.createOne.mockImplementation(async () => createOneResult);
    });

    it('should consume valid request', async () => {
      const requestData = {
        id: '7a570935-3b36-48a2-8889-48ca8810be11',
        manufacturer: {
          id: 'put manufacturer id here'
        },
        price: 100500,
        firstRegistrationDate: '2018-08-16T12:47:15.000Z',
      };

      const { body } = await subject(requestData).expect(201);

      expect(body).toEqual(createOneResult);
      expect(spies.carService.createOne).toHaveBeenCalled();
      expect((spies.carService.createOne.mock.calls as any)[0][1]).toEqual(requestData);
    });

    it('should reject invalid request (no manufacturer)', async () => {
      const { body } = await subject({
        id: '7a570935-3b36-48a2-8889-48ca8810be11',
        price: 100500,
        firstRegistrationDate: '2018-08-16T12:47:15.000Z',
      })
        .expect(400);
      expect(body).toMatchSnapshot();
    });

    it('should reject invalid request (bad price - string)', async () => {
      const { body } = await subject({
        id: '7a570935-3b36-48a2-8889-48ca8810be11',
        manufacturer: {
          id: 'put manufacturer id here'
        },
        price: '100500',
        firstRegistrationDate: '2018-08-16T12:47:15.000Z',
      })
        .expect(400);
      expect(body).toMatchSnapshot();
    });
  });

  it('should fetch car', async () => {
    spies.carService.getOne.mockImplementation(async () => {
      const result = new Car();
      result.price = 100;
      result.id = 'none';
      return result;
    });
    const { body } = await request().get('/car/random-id-value').expect(200);
    expect(body).toMatchSnapshot();
    expect(spies.carService.getOne.mock.calls[0]).toMatchSnapshot();
  });

});
