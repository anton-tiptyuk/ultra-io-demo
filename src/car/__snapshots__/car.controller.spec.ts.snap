// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Car Controller checks for CRUD POST /car should reject invalid request (bad price - string) 1`] = `
Object {
  "error": "Bad Request",
  "message": Array [
    "price must be a number conforming to the specified constraints",
  ],
  "statusCode": 400,
}
`;

exports[`Car Controller checks for CRUD POST /car should reject invalid request (no manufacturer) 1`] = `
Object {
  "error": "Bad Request",
  "message": Array [
    "manufacturer must be an object",
    "manufacturer should not be empty",
  ],
  "statusCode": 400,
}
`;
